hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java

```diff
@@ -699,18 +717,53 @@ private void cleanupMetadataAndUnlock(
     } catch (RuntimeException e) {
       LOG.error("Failed to cleanup metadata file at {}", metadataLocation, e);
     } finally {
-      unlock(lockId);
+      unlock(lockId, agentInfo);
       tableLevelMutex.unlock();
     }
   }
 
-  private void unlock(Optional<Long> lockId) {
-    if (lockId.isPresent()) {
-      try {
-        doUnlock(lockId.get());
-      } catch (Exception e) {
-        LOG.warn("Failed to unlock {}.{}", database, tableName, e);
+  private void unlock(Optional<Long> lockId, String agentInfo) {
+    Long id = null;
+    try {
+      if (!lockId.isPresent()) {
+        // Try to find the lock based on agentInfo. Only works with Hive 2 or later.
+        if (MetastoreClientVersion.min(MetastoreClientVersion.HIVE_2)) {
+          AtomicReference<Long> foundLockId = new AtomicReference<>();
+          AtomicReference<LockState> state = new AtomicReference<>();
+          if (!findLock(agentInfo, foundLockId, state)) {
+            // No lock found
+            LOG.info("No lock found with {} agentInfo", agentInfo);
+            return;
+          }
+
+          id = foundLockId.get();
+        } else {
+          LOG.warn("Could not find lock with HMSClient {}", MetastoreClientVersion.current());
+          return;
+        }
+      } else {
+        id = lockId.get();
+      }
+
+      doUnlock(id);
+    } catch (InterruptedException ie) {
+      if (id != null) {
```

Comment:
Is this just to handle interrupt for unlock?  I'm not sure I see how its related to the rest of the change, I would think if anything here, we should be checking TException for unlock and try unlock again, synonomous with tryLock?

Reply:
TBH this was the original change which started my thinking and exploded into a general hardening 😄 

In Flink we had the following situation:
- Commited a change and tried to unlock the table
- The first unlock failed because of a network issue
- RetryingMetastoreClient started the sleep between retries
- The job was interrupted
- The lock was not removed

So this change aims to try to unlock the table in any case, especially in the situation when there is an interrupt.

Reply:
While this makes sense, following this logic, this change seems like its missing handling for following:

- lock fails due to interrupted (or is this handled?  I didnt see any explicit interruptedException handling in lock code)
- unlock fails because of Thrift Exception

Shouldn't we harden these cases then?

Reply:
If the `tryLock` is failed because of an `InterruptedException` it will throw a `RuntimeException` and does not set the `lockId`:
```
              } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                LOG.warn("Interrupted while acquiring lock on table {}.{}", database, tableName, e);
                throw new RuntimeException("Interrupted while acquiring lock", e);
              }
```

If the `lockId` is not set then we try to use `findLock` to find the lockId, and then `doUnlock` to unlock the table.

We need to harden the HMS `lock` request specifically, because it is not handled by the `RetryingMetastoreClient` because of this annotation:
```
  @RetrySemantics.CannotRetry
  public LockResponse lock(LockRequest rqst) throws NoSuchTxnException, TxnAbortedException, MetaException {
```
For more details see: https://github.com/apache/hive/blob/rel/release-2.3.9/metastore/src/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java#L855-L871

Since both the `unlock`, and the `findLock` is retried, I think we should not do any extra work there, and expect that the `RetryingMetastoreClient` will fix our issues when possible

Reply:
Thanks for the explanation, makes sense.  It does seem a bit risky for complex handling of interrupt by unlocking again, unless you are very sure thats what caused the issue (if the first call is interrupted because of timeout, I'm not sure what the chance is the unlock retry will succeed), what do you think?  Will leave it to you as you have seen the issue.  Also fyi @stevenzwu